File: src/main/nim/eyefuck.nim
# i don't know... nim is, not my favorite (like lua). But i don't hate it (like python)
# it is like... idk 

# there is a problem with it, and i will not fix it!
import os, strutils, strformat, re

const
  Reset = "\e[0m"
  Red = "\e[31m"
  Green = "\e[32m"
  Yellow = "\e[33m"
  Blue = "\e[34m"
  Cyan = "\e[36m"


const
  EyfV = 1.2
  TapeSize = 300000

# ---------------------------
# Eyefuck Interpreter
# ---------------------------
proc runInterpreter(code: string) =
  var tape = newSeq[byte](TapeSize)
  var ptr = 0
  let lines = code.splitLines()
  var loopStack: seq[int] = @[]
  var i = 0

  while i < lines.len:
    var line = lines[i].strip()
    
    # remove comments after #
    let commentPos = line.find('#')
    if commentPos != -1:
      line = line[0..<commentPos].strip()
    
    if line.len == 0:
      inc i
      continue

    case line
    of "^": # increment cell
      inc tape[ptr]
    of "v": # decrement cell
      if tape[ptr] > 0:  # Prevent underflow
        dec tape[ptr]
    of ">": # move pointer right
      ptr = (ptr + 1) mod TapeSize
    of "<": # move pointer left
      ptr = if ptr == 0: TapeSize - 1 else: ptr - 1
    else:
      if line.startsWith("bin"):
        # set cell from binary
        let binStr = line[3..^1].strip()
        try:
          tape[ptr] = byte(parseBinInt(binStr))
        except ValueError:
          stderr.writeLine "Invalid binary format: " & binStr
          return
      elif line.startsWith("col"):
        # set text color from HEX
        let matches = line.findAll(re"\[([0-9A-Fa-f]+)\]")
        if matches.len > 0:
          let hexStr = matches[0][1..^2] # Remove brackets
          try:
            let colorInt = parseHexInt(hexStr)
            let r = (colorInt shr 16) and 0xFF
            let g = (colorInt shr 8) and 0xFF
            let b = colorInt and 0xFF
            stdout.write(fmt"\e[38;2;{r};{g};{b}m")
          except ValueError:
            stderr.writeLine "Invalid HEX color: " & hexStr
            return
      elif line.startsWith("load["):
        # load file
        let matches = line.findAll(re"\[([^\]]+)\]")
        if matches.len > 0:
          let filename = matches[0][1..^2] # Remove brackets
          try:
            let content = readFile(filename)
            # Store first character of file content in current cell
            if content.len > 0:
              tape[ptr] = byte(ord(content[0]))
            else:
              tape[ptr] = 0
          except IOError:
            stderr.writeLine "Error loading file: " & filename
            return
      elif line == ",":
        # read single byte input
        try:
          let input = stdin.readChar()
          tape[ptr] = byte(ord(input))
        except IOError:
          tape[ptr] = 0  # EOF or error case
      elif line == ".":
        # print cell as char
        stdout.write(char(tape[ptr]))
      elif line == "loop[":
        # start loop
        loopStack.add(i)
      elif line == "]":
        # end loop
        if loopStack.len == 0:
          stderr.writeLine "Unmatched ] at line " & $i
          return
        if tape[ptr] != 0:
          i = loopStack[^1]
        else:
          discard loopStack.pop()
      else:
        stderr.writeLine Red & "error caught while parsing"
        stderr.writeLine Red & "at line: " & $i & ": " & line
        return
    
    inc i
  
  stdout.write(Reset)  # Reset color at end
  stdout.writeLine("")  # Ensure newline at end

# ---------------------------
# Interactive REPL
# ---------------------------
proc startREPL() =
  echo Cyan & "Eyefuck DEV 2025 - REPL" & Reset
  echo "Type commands below, empty line to execute, Ctrl+C to exit"
  
  var codeLines: seq[string] = @[]
  
  while true:
    stdout.write("$ ")
    let line = stdin.readLine()
    
    if line.strip().len == 0:
      if codeLines.len > 0:
        runInterpreter(codeLines.join("\n"))
      codeLines = @[]
      continue
    
    codeLines.add(line)

# ---------------------------
# Main function
# ---------------------------
proc main() =
  if paramCount() < 1:
    echo Red & "Usage:" & Reset & " eyefuck <command> [file.eyf]"
    echo "Commands: run, -i, help, about, version, ov"
    quit(1)

  let mode = paramStr(1)

  case mode
  of "run":
    if paramCount() < 2:
      echo Red & "Please specify a file to run." & Reset
      quit(1)
    let file = paramStr(2)
    if not fileExists(file):
      echo "File not found: ", file
      quit(1)
    try:
      let code = readFile(file)
      runInterpreter(code)
    except IOError:
      echo "Error reading file: ", getCurrentExceptionMsg()
      quit(1)
  of "-i", "--i", "i":
    startREPL()
  of "help", "-help", "-h", "--h", "--help":
    echo Cyan & "Eyefuck HELP:" & Reset
    echo Yellow & "  eyefuck run <file.eyf>" & Reset & "  -> " & Green & "execute the Eyefuck file" & Reset
    echo Yellow & "  eyefuck -i" & Reset & "             -> " & Green & "interactive REPL mode" & Reset
    echo Yellow & "  eyefuck about" & Reset & "          -> " & Green & "information about this interpreter" & Reset
    echo Yellow & "  eyefuck version" & Reset & "        -> " & Green & "show version information" & Reset
  of "about":
    echo Cyan & "Eyefuck DEV 2025" & Reset
    echo Green & "MIT license" & Reset & " see LICENSE for more information"
    echo "Please help me motive by giving the repo a star"
    echo Blue & "github:" & Reset & " github.com/bandikaaking"
    echo "crafted with " & Red & "<3" & Reset & " by " & Yellow & "@Bandikaaking" & Reset
  of "version", "--v", "--version", "-v", "v", "-version":
    echo "Current eyefuck version: ", EyfV
  of "ov", "-ov", "--ov":
    echo "Other Eyefuck versions: "
    echo "0.10: Started / added 2 instructions"
    echo "0.11-0.43: Fixed many bugs, and edded 5 more instructions"
    echo "1.0: Added syntax highliting"
    echo "1.1: Fixed bugs"
    echo "added more eyefuck modes / rewrited README.md"
  else:
    echo Red & "Unknown mode:" & Reset, " ", mode
    echo "Use 'eyefuck help' for available commands."
    quit(1)

when isMainModule:
  main()

File: src/main/PHP/eyefuck.php
<?php

/* did you know, i learned php first! */
/* and it's a goddamn plan to write an interpreter in an interpreter! */
// ANSI colors
define('RESET', "\033[0m");
define('RED', "\033[31m");
define('GREEN', "\033[32m");
define('YELLOW', "\033[33m");
define('BLUE', "\033[34m");
define('CYAN', "\033[36m");
define('WHITE', "\033[97m");

define('EYF_V', 1.2);
define('TAPE_SIZE', 300000);

// ---------------------------
// Main function
// ---------------------------
function main() {
    global $argv;
    
    if (count($argv) < 2) {
        echo RED . "Usage:" . RESET . " eyefuck <command> [file.eyf]\n";
        return;
    }

    $mode = $argv[1];

    switch ($mode) {
        case "run":
            if (count($argv) < 3) {
                echo RED . "Please specify a file to run." . RESET . "\n";
                return;
            }
            $file = $argv[2];
            if (!file_exists($file)) {
                echo "Error reading file\n";
                return;
            }
            $code = file_get_contents($file);
            run_interpreter($code);
            break;
            
        case "-i":
        case "--i":
        case "i":
            start_repl();
            break;
            
        case "help":
        case "-help":
        case "-h":
        case "--h":
        case "--help":
            echo CYAN . "Eyefuck HELP:" . RESET . "\n";
            echo YELLOW . "  eyefuck run <file.eyf>" . RESET . "  -> " . GREEN . "execute the Eyefuck file" . RESET . "\n";
            echo YELLOW . "  eyefuck -i" . RESET . "             -> " . GREEN . "interactive REPL mode" . RESET . "\n";
            echo YELLOW . "  eyefuck about" . RESET . "          -> " . GREEN . "information about this interpreter" . RESET . "\n";
            break;
            
        case "about":
            echo CYAN . "Eyefuck DEV 2025" . RESET . "\n";
            echo GREEN . "MIT license" . RESET . " see LICENSE for more information\n";
            echo "Please help me motive by giving the repo a star\n";
            echo BLUE . "github:" . RESET . " github.com/bandikaaking\n";
            echo "crafted with " . RED . "<3" . RESET . " by " . YELLOW . "@Bandikaaking" . RESET . "\n";
            break;
            
        case "version":
        case "--v":
        case "--version":
        case "-v":
        case "v":
        case "-version":
            echo "Current eyefuck version: " . EYF_V . "\n";
            break;
            
        case "ov":
        case "-ov":
        case "--ov":
            echo "Other Eyefuck versions: \n";
            echo "0.10: Started / added 2 instructions\n";
            echo "0.11-0.43: Fixed many bugs, and edded 5 more instructions\n";
            echo "1.0: Added syntax highliting\n";
            echo "1.1: Fixed bugs\n";
            echo "added more eyefuck modes / rewrited README.md\n";
            break;
            
        default:
            echo RED . "Unknown mode:" . RESET . " " . $mode . "\n";
            break;
    }
}

// ---------------------------
// Interactive REPL
// ---------------------------
function start_repl() {
    echo CYAN . "Eyefuck DEV 2025 - REPL" . RESET . "\n";
    echo "Type commands below, empty line to execute, Ctrl+C to exit\n";
    
    $code_lines = array();
    
    while (true) {
        echo "$ ";
        $line = trim(fgets(STDIN));
        
        if ($line === "") {
            run_interpreter(implode("\n", $code_lines));
            $code_lines = array();
            continue;
        }
        
        $code_lines[] = $line;
    }
}

// ---------------------------
// Eyefuck Interpreter
// ---------------------------
function run_interpreter($code) {
    $tape = array_fill(0, TAPE_SIZE, 0);
    $ptr = 0;
    $lines = explode("\n", $code);
    $loop_stack = array();
    
    for ($i = 0; $i < count($lines); $i++) {
        $line = trim($lines[$i]);
        
        // remove comments after #
        $comment_pos = strpos($line, '#');
        if ($comment_pos !== false) {
            $line = trim(substr($line, 0, $comment_pos));
        }
        
        if ($line === "") {
            continue;
        }

        switch ($line) {
            case "^": // increment cell
                $tape[$ptr] = ($tape[$ptr] + 1) % 256;
                break;
                
            case "v": // decrement cell
                $tape[$ptr] = ($tape[$ptr] - 1) % 256;
                break;
                
            case ">": // move pointer right
                $ptr = ($ptr + 1) % TAPE_SIZE;
                break;
                
            case "<": // move pointer left
                $ptr = $ptr == 0 ? TAPE_SIZE - 1 : $ptr - 1;
                break;
                
            default:
                if (strpos($line, "bin") === 0) {
                    // set cell from binary
                    $bin = trim(substr($line, 3));
                    $tape[$ptr] = bindec($bin) % 256;
                }
                elseif (strpos($line, "col") === 0) {
                    // set text color from HEX
                    if (preg_match('/\[([0-9A-Fa-f]+)\]/', $line, $matches)) {
                        $hex = $matches[1];
                        $color_int = hexdec($hex);
                        $r = ($color_int >> 16) & 0xFF;
                        $g = ($color_int >> 8) & 0xFF;
                        $b = $color_int & 0xFF;
                        echo "\033[38;2;{$r};{$g};{$b}m";
                    }
                }
                elseif (strpos($line, "load[") === 0) {
                    // load file
                    if (preg_match('/\[([^\]]+)\]/', $line, $matches)) {
                        $filename = $matches[1];
                        if (file_exists($filename)) {
                            $tape[$ptr] = 0;
                        }
                    }
                }
                elseif ($line === ",") {
                    // read single byte input
                    $input = fgetc(STDIN);
                    if ($input !== false) {
                        $tape[$ptr] = ord($input);
                    }
                }
                elseif ($line === ".") {
                    // print cell as char
                    echo chr($tape[$ptr]);
                }
                elseif ($line === "loop[") {
                    // start loop
                    array_push($loop_stack, $i);
                }
                elseif ($line === "]") {
                    // end loop
                    if ($tape[$ptr] != 0) {
                        if (!empty($loop_stack)) {
                            $i = end($loop_stack);
                        } else {
                            echo "Unmatched ]\n";
                            return;
                        }
                    } else {
                        if (!empty($loop_stack)) {
                            array_pop($loop_stack);
                        }
                    }
                }
                else {
                    echo RED . "error caught while parsing\n";
                    echo RED . "at line: " . $line . RESET . "\n";
                    return;
                }
                break;
        }
    }
    
    echo "\n";
}

// Run main if this file is executed directly
if (isset($argv) && count($argv) > 0 && basename($argv[0]) == basename(__FILE__)) {
    main();
}
?>